certbot/tests/client_test.py

```diff
@@ -214,12 +289,72 @@ def test_obtain_certificate(self, mock_crypto_util):
             self.client.obtain_certificate(domains),
             (mock.sentinel.certr, mock.sentinel.chain, mock.sentinel.key, csr))
 
-        mock_crypto_util.init_save_key.assert_called_once_with(
-            self.config.rsa_key_size, self.config.key_dir)
+        mock_crypto_util.save_key.assert_called_once_with(
+            mock_crypto_util.make_key_ecdsa(), self.config.key_dir)
         mock_crypto_util.init_save_csr.assert_called_once_with(
             mock.sentinel.key, domains, self.config.csr_dir)
         self._check_obtain_certificate()
 
+    @mock.patch("certbot.client.crypto_util")
+    def test_obtain_certificate_multi_alg(self, mock_crypto_util):
+        self._mock_obtain_certificate()
+
+        self.config.key_types = "ecdsa rsa"
+        csr = le_util.CSR(form="der", file=None, data=CSR_SAN)
+        mock_crypto_util.init_save_csr.return_value = csr
+        mock_crypto_util.save_key.return_value = mock.sentinel.key
+        domains = ["example.com", "www.example.com"]
+
+        # return_value is essentially set to (None, None) in
+        # _mock_obtain_certificate(), which breaks this test.
+        # Thus fixed by the next line.
+
+        authzr = []
+
+        for domain in domains:
+            authzr.append(
+                mock.MagicMock(
+                    body=mock.MagicMock(
+                        identifier=mock.MagicMock(
+                            value=domain))))
+
+        self.client.auth_handler.get_authorizations.return_value = authzr
+
+        self.assertRaises(errors.Error,
+            self.client.obtain_certificate(domains),
+            (mock.sentinel.certr, mock.sentinel.chain, mock.sentinel.key, csr))
```

Comment:
Here you're calling `self.client.obtain_certificate(domains)`, and then passing the result to `self.assertRaises()`. If `obtain_certificate()` raises an exception, which it does, then `self.assertRaises()` will never be called.

You want
```python
with self.assertRaises(errors.Error):
    self.client.obtain_certificate(domains)
```
or
```
self.assertRaises(errors.Error, self.client.obtain_certificate, domains)
```

I'm not sure what `(mock.sentinel.certr, ...)` are supposed to be -- the assertion message?  You cannot pass a custom assertion message to `assertRaises`, the API [doesn't allow that](https://docs.python.org/2.7/library/unittest.html#unittest.TestCase.assertRaises).